### ISSUE TITLE
Fix `torch.stft` and `torch.istft` when using `center=False`, non-rectangular windows and `win_length==hop_length`

### ISSUE THREAD

**Issue by philgzl** [pytorch/pytorch#134323]
### 🚀 The feature, motivation and pitch

Over the past years, multiple issues were opened about using the non-default parameters in `torch.stft` and `torch.istft`. Those issues have not been solved and are still crippling. This is significant as the parameters that can raise errors are commonly used and well motivated in audio signal processing.

This is a non-exhaustive compilation of issues:
- pytorch/audio/issues/427
- pytorch/audio/issues/500
- #62323
- #81428
- #91309
- #118507
- #70073

The importance of those issues is also reflected by the number of re-implementations of iSTFT in audio signal processing repositories (e.g. [this one](https://github.com/charactr-platform/vocos/blob/03c4fcbb321e4b04dd9b5091486eedabf1dc9de0/vocos/spectral_ops.py#L7-L75) by @hubertsiuzdak)

Those issues most commonly happen when setting `center=False` as the NOLA check fails at the edges, or when using windows with zeroes like Hann or the ones proposed by Mauler and Martin in [2007](https://ieeexplore.ieee.org/document/7098797) and [2009](https://doi.org/10.1155/2009/469480) as illustrated by @LXP-Never [here](https://github.com/pytorch/pytorch/issues/62323#issuecomment-1829564313).

The solutions are:
- Adding an option to disable the NOLA check
- Adding support for custom padding
- Fixing the normalization by taking the selected window into account when `normalized=True`.

In my opinion, the current padding is fundamentally wrong and should be reimplemented. As pointed out by @fakufaku [here](https://github.com/pytorch/pytorch/issues/91309#issuecomment-1371755324), the option `center=True` currently does two unrelated things, 1. fitting an integer number of frames and 2. ensuring the signal is invertible by dealing with the first and last samples when using a non-rectangular window, in which case the padding should actually be `win_length - hop_length`, and not `win_length // 2`. As I demonstrated [here](https://github.com/pytorch/pytorch/issues/70073#issuecomment-1495893892), the current padding can cause data loss even when `center=True`!

**Comment by albanD** [comment]
cc @ezyang as the last person who reviewed a PR changing this function

**Comment by ezyang** [comment]
Reviewing the issues in question, it seems clear that we are lacking a dedicated maintainer for this cluster of functionality (@peterbell10 and @nateanl, would you agree?) So if someone does want to push forward with a plan, I would love to facilitate them. @philgzl, are you this person?

In general, adding more options to extend the functionality of stft/istft is not a big deal, if you give a plausible proposal we can accept it. Changing the normalization behavior is trickier, as it may potentially be a load bearing bug that would affect the reproducibility of other models. As a non-expert, I don't have a good sense for whether or not it is worth BC-breaking it, but I am happy to entertain it as a possibility, esp if you're willing to keep the old buggy padding around as an option.

**Comment by philgzl** [comment]
> So if someone does want to push forward with a plan, I would love to facilitate them. @philgzl, are you this person?

What do you mean exactly by "pushing forward with a plan"? Opening a pull request? I can try to give it a shot when I have the time.

> Changing the normalization behavior is trickier, as it may potentially be a load bearing bug that would affect the reproducibility of other models. As a non-expert, I don't have a good sense for whether or not it is worth BC-breaking it, but I am happy to entertain it as a possibility, esp if you're willing to keep the old buggy padding around as an option.

I think the normalization is minor compared to the padding issue. If changing it is tricky, a note in the documentation emphasizing that the normalization does not match common L2 normalization/does not satisfy Parseval's identity would already be valuable IMO.

Padding could be solved by adding a `padding` option which could be `None` (current behavior), `'minimum'` (i.e. `win_length - hop_length`), or an integer (or a callable?) for custom padding. This could be implemented while keeping the `center` option, but `center=True` and `padding!=None` would be incompatible. IMO the `minimum` padding should be the default behavior (I don't get the motivation for having the first frame "centered in time"), and it would already fix NOLA errors when the padding caused by `center=True` is not enough. Note that we might still need to implement additional padding at the end of the signal to ensure trailing samples are not discarded as illustrated [here](https://github.com/pytorch/pytorch/issues/70073#issuecomment-1495893892).

Finally a `disable_nola_check` option in `istft` could be added to normalize the output with the window envelope only in the regions where the window envelope is non-zero as suggested in [this thread](https://github.com/pytorch/pytorch/issues/91309). This logic would match the default librosa and MATLAB implementation.

**Comment by ezyang** [comment]
> What do you mean exactly by "pushing forward with a plan"? Opening a pull request? I can try to give it a shot when I have the time.

Yep, designing out the full set of changes, and then opening PRs / reviewing PRs that move us toward it.

> Padding could be solved by adding a padding option which could be None (current behavior), 'minimum' (i.e. win_length - hop_length), or an integer (or a callable?) for custom padding.

In eager mode, a callable for custom padding will be difficult to implement efficiently. If someone can cough up a Triton implementation of istft we could do something similar to flex attention to allow for custom padding.  If you want to support both minimum and integer paddings, you'll need two separate mutually exclusive new args as we don't support Union in native_functions.yaml.

**Comment by ezyang** [comment]
@philgzl are you willing to help me check if the change in https://github.com/pytorch/pytorch/pull/143142 is the same thing as here? (I had someone else propose this change to stft but I am not an FFT expert so I can't tell if this is the same as the problem you described here.)

**Comment by philgzl** [comment]
- The first change from `n_frames = 1 + (len - n_fft) / hop_length` to `n_frames = 1 + (len - win_length) / hop_length` makes sense. The number of frames should not depend on `n_fft`. `n_fft` should only result in interpolation in the frequency domain.
- I am not sure I understand the second change `input = at::pad(input, {(n_fft - win_length) / 2, (n_fft - win_length) / 2});`. Is it implementing the time-domain padding required to fit the new `n_frames` of size `n_fft` with hop size `hop_length`? If so this feels hacky and confusing because `n_frames` is calculated as if the window size is `win_length`, but the striding operation still uses a size of `n_fft`. Also I suspect that this new padding causes frames to not be centered at t*hop_length when `center=True` as the documentation wants it. Unless you consider the frame size is `n_fft` instead of `win_length`, but IMO that's counter intuitive.

IMO the correct way is to frame/stride/unfold with a size of `win_length`, multiply each frame by the window without padding it to `n_fft`, and finally pad each frame to `n_fft` (to the right) right before computing the FFT for frequency interpolation. Before framing, pad left and right with `win_length - hop_length` if perfect reconstruction is wanted. `center` should be deprecated.

It's not entirely clear to me if the proposed changes will fix all the issues from my original post.

**Comment by philgzl** [comment]
Sharing [my STFT implementation](https://gist.github.com/philgzl/4698258da4f074ce8cd387da40d86e56) here to illustrate my points and for future readers looking for solutions.

**Comment by dvorjackz** [comment]
@philgzl just for some context, we are considering adding a feature flagged version of https://github.com/pytorch/pytorch/pull/143142 for internal reasons that are entirely separate from this post. This is to cover the case when center = false and win_length < n_ftt, which would prevent the start of the signal from `[0:(n_fft - win_length)/2]` from being dropped when there is no padding (which doesn't seem to work with center = false anyway).

As for the issues in this post, I believe the simplest solution that also preserves BC would be to add an option to disable the NOLA check and do something like this https://github.com/pytorch/pytorch/issues/91309#issuecomment-1364021931, along with a better warning message and documentation. What are your thoughts?

**Comment by philgzl** [comment]
Yes sorry for digressing. There seem to be two separate problems, 1. the framing operation and 2. the NOLA check.

> but are you sure this makes it broken?

In my opinion, yes. See [here](https://github.com/pytorch/pytorch/issues/70073#issuecomment-1495893892), [here](https://github.com/pytorch/pytorch/issues/142392), your own issue described above, and my code snippet above. I stopped using `stft` for my projects and instead use `unfold` + `rfft`.

> Librosa also [pads the window left and right to match n_fft](https://github.com/librosa/librosa/blob/ebd878fabb87c588f98f372a94fca5d167c59cf2/librosa/core/spectrum.py#L244) before unfolding, and it [unfolds with a kernel size of n_fft as well](https://github.com/librosa/librosa/blob/ebd878fabb87c588f98f372a94fca5d167c59cf2/librosa/core/spectrum.py#L343).

MATLAB and [scipy.signal.stft](https://github.com/scipy/scipy/blob/6e246d0b54dd55dc69232a0caae6772228a7ac25/scipy/signal/_spectral_py.py#L2158C1-L2158C71) pad only upon calculating the FFT. I also think it's weird to have separate options for the frame length and the number of FFT points, but use the number of FFT points as the frame length internally.

The way I have seen STFT taught is first with a frame length and a hop length parameter. The number of FFT points parameter comes only after in case you want to artificially increase the frequency resolution by padding each frame. But that is independent of the framing operation.

It could be that #143142 actually makes the two approaches equivalent and fixes this.

> To address just the theme of the issues linked here, a NOLA opt-out should do the job

Yes, an option to disable the NOLA check would solve most of the issues linked in the original post.

**Comment by ezyang** [comment]
> MATLAB and [scipy.signal.stft](https://github.com/scipy/scipy/blob/6e246d0b54dd55dc69232a0caae6772228a7ac25/scipy/signal/_spectral_py.py#L2158C1-L2158C71) pad only upon calculating the FFT. I also think it's weird to have separate options for the frame length and the number of FFT points, but use the number of FFT points as the frame length internally.

Just to make sure I understand, your claim is that librosa also does it wrong, is that right?

**Comment by dvorjackz** [comment]
@philgzl I think we are on the same page actually! On the topic of framing, the proposed change in https://github.com/pytorch/pytorch/pull/143142 should address your concerns. I'm unable to speak as to whether padding should go before or after FFT, but the effect of this change will be the same, since to compensate for padding the window, we additionally pad the input signal by the same amount so that the window starts at t = 0 of the original input signal. Then in theory this is the same as unfolding with a kernel size of win_length instead of n_fft.

The change will allow us to relax the check [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/SpectralOps.cpp#L910), and for your provided example

> import torch
> 
> y = torch.stft(
>     torch.randn(1000),
>     win_length=100,
>     hop_length=100,
>     n_fft=1024,
>     center=False,
>     return_complex=True,
> )

It will end up fitting exactly 10 frames. We are planning on gating this change behind a optional parameter to preserve BC.

**Comment by philgzl** [comment]
> Just to make sure I understand, your claim is that librosa also does it wrong, is that right?

Let's see.

```python
import librosa
import numpy as np

# check case where n_fft bigger than input length
y = librosa.stft(
    np.random.randn(1000),
    win_length=100,
    hop_length=100,
    n_fft=1024,
    center=False,
)  # raises an error

# check if first samples are discarded
x = np.zeros(1000)
x[0] = 1
y = librosa.stft(
    x,
    win_length=100,
    hop_length=100,
    n_fft=200,
    center=False,
    window="boxcar",
)
assert not (y==0).all()  # fails
```
Seems like librosa also uses a `n_fft`-dependent framing. It crashes when n_fft is bigger than input length and it discards the first samples. So yes, I believe librosa does it wrong.

> https://github.com/pytorch/pytorch/pull/143142 should address your concerns

I agree it sounds like it could be equivalent. I would have to sit with pen and paper to be convinced of it. But if I quickly test with the code snippet below it seems like it's not equivalent (built this very quickly, didn't investigate further).

```python
import torch
import torch.nn.functional as F

win_length = 512
hop_length = 256
n_fft = 1024
window = torch.ones(win_length)

def pad_after(x):
    x = x.unfold(0, size=win_length, step=hop_length)
    x = torch.fft.rfft(x * window, n=n_fft)
    return x

def pad_before(x):
    padding = (n_fft - win_length) // 2
    window_ = F.pad(window, (padding, padding))
    x = F.pad(x, (padding, padding))
    x = x.unfold(0, size=n_fft, step=hop_length)
    x = torch.fft.rfft(x * window_)
    return x

x = torch.sin(torch.arange(4096) * 0.01)

y1 = pad_after(x)
y2 = pad_before(x)

assert y1.shape == y2.shape
assert torch.allclose(y1, y2)  # fails
````

**Comment by philgzl** [comment]
So after thinking about it a bit longer, I am convinced that the two approaches (torch with #143142 vs. MATLAB/scipy/my approach) are not equivalent. The difference is in the phase, since left-padding the input and the window introduces group delay. This is only caused by the left-padding, not by the striding with `size=n_fft`. If torch was right-padding with `n_fft-win_length` instead of left- and right-padding with `(n_fft-win_length)//2`, the two approaches would be equivalent.

Below is an example showing this. The input is a dirac impulse at t=0. The input length is equal to the frame length, so the STFT output has one frame, whose magnitude and phase I am plotting. The magnitude should be flat (all frequencies are in the input), and the phase should be all-zero (all frequencies are at t=0). [Increasing `n_fft` should only do interpolation](https://ccrma.stanford.edu/~jos/st/Ideal_Spectral_Interpolation.html). As you can see, with the torch approach (padding signal and window left and right), increasing `n_fft` correctly interpolates the magnitude (still flat), but does not do so for the phase (not all-zero anymore). Basically the torch approach is correctly indicating there is a dirac impulse, but is wrongly indicating it is delayed (negative linear slope means group delay).


```python
import matplotlib.pyplot as plt
import numpy as np
import torch
import torch.nn.functional as F

win_length = 512
hop_length = 256
window = torch.ones(win_length)

def matlab_scipy_stft(x, n_fft):
    x = x.unfold(0, size=win_length, step=hop_length)
    x = torch.fft.rfft(x * window, n=n_fft)
    return x

def torch_stft(x, n_fft):
    padding = (n_fft - win_length) // 2
    window_ = F.pad(window, (padding, padding))
    x = F.pad(x, (padding, padding))
    x = x.unfold(0, size=n_fft, step=hop_length)
    x = torch.fft.rfft(x * window_)
    return x

x = torch.zeros(512)
x[0] = 1

f1 = np.arange(257) / 512
f2 = np.arange(513) / 1024

fig, ax = plt.subplots(2, 2)
ax[0, 0].plot(f1, matlab_scipy_stft(x, 512)[0].abs(), label="n_fft=512")
ax[0, 0].plot(f2, matlab_scipy_stft(x, 1024)[0].abs(), label="n_fft=1024")
ax[1, 0].plot(f1, np.unwrap(matlab_scipy_stft(x, 512)[0].angle()), label="n_fft=512")
ax[1, 0].plot(f2, np.unwrap(matlab_scipy_stft(x, 1024)[0].angle()), label="n_fft=1024")
ax[0, 1].plot(f1, torch_stft(x, 512)[0].abs(), label="n_fft=512")
ax[0, 1].plot(f2, torch_stft(x, 1024)[0].abs(), label="n_fft=1024")
ax[1, 1].plot(f1, np.unwrap(torch_stft(x, 512)[0].angle()), label="n_fft=512")
ax[1, 1].plot(f2, np.unwrap(torch_stft(x, 1024)[0].angle()), label="n_fft=1024")
ax[0, 0].title.set_text("Magnitude (MATLAB/scipy)")
ax[1, 0].title.set_text("Phase (MATLAB/scipy)")
ax[0, 1].title.set_text("Magnitude (torch)")
ax[1, 1].title.set_text("Phase (torch)")
for ax_ in ax.flatten():
    ax_.legend()
fig.tight_layout()
```

![Image](https://github.com/user-attachments/assets/dbf62096-9489-417a-84ff-ddf1739dcc5a)

TLDR; with #143142 we get the correct number of frames and the correct magnitude. But when increasing `n_fft`, we not only interpolate in the frequency domain, we also introduce additional group delay. If you ask a DSP person, they will tell you that's wrong, and that you should pad right only.

**Comment by dvorjackz** [comment]
Hi @philgzl, makes sense - seems then the difference is in how we are padding the window (`window` or `win_length`) to fill the frame (`n_fft`).

Then https://github.com/pytorch/pytorch/pull/145324 doesn't actually address your specific issue (since it actually doesn't matter whether you unwrap with a kernel size of win_length or n_fft, it just matters what the hop_length is and how the input is padded). However, it is a change that we need to make for internal reasons (since it allows us to center without dropping initial frames when `win_length < n_fft`) and that has been discussed in librosa before.

That being said, if I'm understanding correctly, what we really need to change is the window padding, not the input signal padding. So would it work if we added the option to pad the window right instead of center to fill the frame? A string parameter specifying how the window is padded which defaults to "center" to preserve backwards compatibility would do the job while also allowing us to explore more window padding options in the future. Then we could use something like `center = False, <window-pad-parameter-name> = "left"` along with manually padding the end of the input signal before calling stft to achieve your use case.

Side note - for the example you mentioned here:

> import torch
> import torch.nn.functional as F
> 
> win_length = 512
> hop_length = 256
> n_fft = 1024
> window = torch.ones(win_length)
> 
> def pad_after(x):
>     x = x.unfold(0, size=win_length, step=hop_length)
>     x = torch.fft.rfft(x * window, n=n_fft)
>     return x
> 
> def pad_before(x):
>     padding = (n_fft - win_length) // 2
>     window_ = F.pad(window, (padding, padding))
>     x = F.pad(x, (padding, padding))
>     x = x.unfold(0, size=n_fft, step=hop_length)
>     x = torch.fft.rfft(x * window_)
>     return x
> 
> x = torch.sin(torch.arange(4096) * 0.01)
> 
> y1 = pad_after(x)
> y2 = pad_before(x)
> 
> assert y1.shape == y2.shape
> assert torch.allclose(y1, y2)  # fails

I'm not too concerned about the differences since`torch.rfft` actually [does some padding before the fft](https://pytorch.org/docs/stable/generated/torch.fft.rfft.html), and this padding around the signal [doesn't seem to be symmetric](https://github.com/pytorch/pytorch/blob/bfaf76bfc677abff6e2f03925c3d1223068d660f/torch/_refs/fft.py#L101-L109). So that's probably where the difference in output arises.

**Comment by DietBru** [comment]
I'd like to make an unsolicited remark, if I may (I landed here by commenting on #145324):

To obtain consistent behavior of the STFT, the input signal should be zero-padded on _both_ sides to include all time slices which (also partly) touch the signal.

The argument goes like this:
Since the STFT is a linear operator, it is not unreasonable to expect to be able to slice a long signal into chunks, calculate the STFT of each chunk and then assemble those STFT chunks together to obtain the STFT of the complete signal. 
When assembling those STFT chunks, the temporal overlapping of two consecutive chunks needs to be taken into account—a simple example is given on the [ShortTimeFFT] page. Hence, when slicing an STFT into two, there is a "leaking" in the direction of the time axis.

On a side note: The fact that the time domain of the STFT is (almost always) larger than that of the input signal may be interpreted as a manifestation of a  phenomenon often referred to as time-frequency [Heisenberg uncertainty principle]. 


[ShortTimeFFT]: https://scipy.github.io/devdocs/reference/generated/scipy.signal.ShortTimeFFT.html
[Heisenberg uncertainty principle]: https://en.wikipedia.org/wiki/Uncertainty_principle#Signal_processing
